### PR TITLE
feat: redundant rpcs

### LINKED ISSUE #1198: Allow multiple RPC's to be defined
I'd love to see something like this....

```
  osmosis:
    type: cosmos
    value:
      key: default
      chain-id: osmosis-1
      rpc-addr: [
        https://rpc.osmosis.strange.love:443,
        https://rpc.osmosis.strange.love:443
     ]
```

Maybe start simple like, is the first item in the list until it gets `5` non `HTTP 200` codes returned, then it tries the next item in the list.

End goal could be, doing health checks against [chain-registry](https://github.com/cosmos/chain-registry) to determine live peers.

Tasks 
- [x] Create fallback-rpc-addrs command to allow multiple comma separated urls
- [x] Add subcommand to allow adding urls from a file that contains said comma separated url. 
- [ ] Add failsafe to continue to next available url after 5 retires . 
```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by joelsmith-2019**
Closes #1198 

**Adds**
- New config field `backup-rpc-addrs` to the chain provider configuration
- CMD `rly chains set-backup-rpc-addrs` to update config.yaml with backup rpcs
- Go routine per chain to check for liveliness (every 10s) & cycles when rpc does not respond
- When adding a chain via the CLI, **2** backup RPCs will be selected from cosmos.directory and defined in the config

**Note**
- When RPCs are cycled, the primary takes precedence. Additionally, the backup rpcs will be tried in the order they are configured.  
- When loading a chain configuration from file, `backup-rpc-addrs` will be picked up as well. Tested locally.

### CODE REVIEW

File: cmd/chains.go
Comment (reecepbcups) [inline cmd/chains.go:90]: note: are we go 1.22?

File: cregistry/chain_info.go
Comment (reecepbcups) [inline cregistry/chain_info.go:311]: why only 2 as backup? should they be able to config this? or use all by default? whats the downside to allowing all if they can set more than 2?

File: cregistry/chain_info.go
Comment (reecepbcups) [inline cregistry/chain_info.go:214]: nit: `forceAdd` name confuses me a little bit as nothing is being added. Maybe rename to `healthyOnly` or `forceUseAll`? (same for the flag --force-add)

File: interchaintest/backup_rpc_test.go
Comment (reecepbcups) [inline interchaintest/backup_rpc_test.go]: nit: I think there is a poll for ack interchaintest function which waits for a successful IBC packet (vs just arb waiting 20 blocks). Not required but could speed this test up by ~30 seconds

File: relayer/chain.go
Comment (reecepbcups) [inline relayer/chain.go:34]: is this state breaking for current relayer configs or is cc.PCfg used?

File: relayer/chains/cosmos/provider.go
Comment (reecepbcups) [inline relayer/chains/cosmos/provider.go:329]: In this hypothetical:
- We are relaying a packet using RPCAddr
- This go routine runs & then waits 10s
- The packet relaying from RPCAddr goes down mid flight
- We wait 10s for the liveness check to set the new RPC
- it finishes the IBC lifecycle

Does it make sense to just roundrobin live backups only when the main RPC goes down & we get errors? Maybe this is an non issue but would give a faster turn around time without the extra status request checks always

i.e. be more reactive to downtime vs proactive to save infra request. May have simpler logic as well?

File: cmd/chains.go
Comment (reecepbcups) [reply cmd/chains.go:90]: we are not, should bump to go1.22 or do the `rpc_address := rpc_address` thing

File: relayer/chain.go
Comment (joelsmith-2019) [reply relayer/chain.go:34]: That field was nowhere to be used throughout the entire project. I'm not sure why it was there to begin with. I removed it to clear things up regarding where the RPCs should reside.

File: interchaintest/backup_rpc_test.go
Comment (joelsmith-2019) [reply interchaintest/backup_rpc_test.go]: Oh right! I'll make sure to swap that out so we can speed up the test. Thanks!

File: cregistry/chain_info.go
Comment (joelsmith-2019) [reply cregistry/chain_info.go:311]: I was originally told to just have it set 2 backup rpcs. There isn't a downside to having more than 2.

File: cmd/chains.go
Comment (joelsmith-2019) [reply cmd/chains.go:90]: I was under the impression that redefining the loop variables was only necessary when calling go routines with those variables. I'll definitely add that additional check to ensure we don't run into that issue, though.

File: relayer/chains/cosmos/provider.go
Comment (joelsmith-2019) [reply relayer/chains/cosmos/provider.go:329]: That's all correct. Right now this implementation is very proactive about swapping RPCs as needed. I did discuss with Justin about making a more responsive approach, so I'll investigate that some more. This initial design was used in the hopes I would not have to inject wrapper logic on all of the methods using the RPCClient.

File: cregistry/chain_info.go
Comment (joelsmith-2019) [reply cregistry/chain_info.go:214]: I agree, the forceAdd is a little vague and misleading. I was trying to keep the same lingo going that we've previously had & I wanted it to match the flag.

@jtieri do you care if I swap out `forceAdd` (including the flag)? I know we discussed making this feature have no "breaking" changes.

File: relayer/chains/cosmos/provider.go
Comment (joelsmith-2019) [reply relayer/chains/cosmos/provider.go:329]: I've done some investigation and it's looking like we'd need to inject logic to swap RPCs in at least 5 different places per provider. It definitely makes the code messier and more difficult to maintain IMO. 

The current approach will hit the nodes a bit more than they normally do, but I find it is much easier to manage as all of the RPC rotation code is centralized in one area.

Comment (reecepbcups) [reply relayer/chains/cosmos/provider.go:329]: makes sense!

File: cmd/chains.go
Comment (reecepbcups) [reply cmd/chains.go:90]: yea even the Go devs have messed up a bunch with no putting it where it needs to go in the main codebase. I can never remember the edge cases so always do it for safety (until 1.22 then we can remove w/ linter)

**Review by reecepbcups — APPROVED** [review]
The [`forceAdd`](https://github.com/cosmos/relayer/pull/1467#discussion_r1635099398) nit is not as important. Approving so it's not blocked on something small

File: cregistry/chain_info.go
Comment (jtieri) [reply cregistry/chain_info.go:214]: i'm indifferent on this change. if you want to go ahead and change things to be consistent you could, we will just have to rethink what we end up tagging the next release as

Comment (joelsmith-2019) [reply cregistry/chain_info.go:214]: Sounds good, I'm going to leave that for another PR as it seems out of place here.

**Pull request merged**
